tests/src/test/kotlin/se/ansman/kotshi/TestDefaultValues.kt

```diff
@@ -0,0 +1,201 @@
+                v13 = GenericClassWithConstructorAsDefault("v13"),
+                v14 = 14)
+
+        expected.testFormatting(json)
+    }
+
+    @Test
+    fun withNullValues() {
+        val expected = ClassWithDefaultValues(
+                v1 = WithCompanionFunction("WithCompanionFunction"),
+                v2 = WithStaticFunction("WithStaticFunction"),
+                v3 = WithCompanionProperty("WithCompanionProperty"),
+                v4 = WithStaticProperty("WithStaticProperty"),
+                v5 = GenericClassWithDefault(null),
+                v6 = GenericClassWithDefault(4711),
+                v7 = LocalDate.MIN,
+                v8 = LocalTime.MIN,
+                v9 = LocalDateTime.MIN,
+                v10 = WithCompanionFunction("v10"),
+                v11 = WithCompanionFunction("DefaultQualifier"),
+                v12 = ClassWithConstructorAsDefault("ClassWithConstructorAsDefault"),
+                v13 = GenericClassWithConstructorAsDefault(null),
+                v14 = 4711)
+
+        val actual = moshi.adapter(ClassWithDefaultValues::class.java).fromJson("""{
+             |  "v1": null,
+             |  "v2": null,
+             |  "v3": null,
+             |  "v4": null,
+             |  "v5": null,
+             |  "v6": null,
+             |  "v7": null,
+             |  "v8": null,
+             |  "v9": null,
+             |  "v10": {
+             |    "v": "v10"
+             |  },
+             |  "v11": null,
+             |  "v12": null,
+             |  "v13": null,
+             |  "v14": null
+             |}""".trimMargin())
+
+        assertEquals(expected, actual)
+    }
+
+    @Test
+    fun withAbsentValues() {
+        val expected = ClassWithDefaultValues(
+                v1 = WithCompanionFunction("WithCompanionFunction"),
+                v2 = WithStaticFunction("WithStaticFunction"),
+                v3 = WithCompanionProperty("WithCompanionProperty"),
+                v4 = WithStaticProperty("WithStaticProperty"),
+                v5 = GenericClassWithDefault(null),
+                v6 = GenericClassWithDefault(4711),
+                v7 = LocalDate.MIN,
+                v8 = LocalTime.MIN,
+                v9 = LocalDateTime.MIN,
+                v10 = WithCompanionFunction("v10"),
+                v11 = WithCompanionFunction("DefaultQualifier"),
+                v12 = ClassWithConstructorAsDefault("ClassWithConstructorAsDefault"),
+                v13 = GenericClassWithConstructorAsDefault(null),
+                v14 = 4711)
+
+        val actual = moshi.adapter(ClassWithDefaultValues::class.java).fromJson("""{
+             |  "v10": {
+             |    "v": "v10"
+             |  }
+        |}""".trimMargin())
+        assertEquals(expected, actual)
+    }
+
+    @Test
+    fun throwsNPEWhenNotUsingDefaultValues() {
+        try {
+            moshi.adapter(ClassWithDefaultValues::class.java).fromJson("{}")
+            fail()
+        } catch (e: NullPointerException) {}
```

Comment:
Worth asserting the message contents here? E.g. to ensure it complains about the right field(s) and that it's not an "unexpected" NPE?